File: backend/src/models/booking.model.ts
import mongoose from "mongoose";

export type PaymentMethod = "payhere" | "cash" | "pay_on_arrival";

export interface IBooking {
  userId: mongoose.Types.ObjectId;
  status: "pending" | "confirmed" | "completed" | "cancelled" | "failed";
  paymentMethod: PaymentMethod;
  paymentStatus: "pending" | "paid" | "unpaid" | "failed";
  totalAmount: number;
  items: {
    serviceId: mongoose.Types.ObjectId;
    selectedDate: Date;
    timeSlotStart: string;
    timeSlotEnd: string;
    quantity: number;
    priceAtBooking: number;
  }[];
  statusHistory: {
    status: string;
    timestamp: Date;
    reason?: string;
  }[];
  cancelledAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const bookingSchema = new mongoose.Schema<IBooking>(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },
    status: {
      type: String,
      enum: ["pending", "confirmed", "completed", "cancelled", "failed"],
      default: "pending",
    },
    paymentMethod: {
      type: String,
      enum: ["payhere", "cash", "pay_on_arrival"],
      required: true,
    },
    paymentStatus: {
      type: String,
      enum: ["pending", "paid", "unpaid", "failed"],
      default: "pending",
    },
    totalAmount: { type: Number, required: true, min: 0 },
    items: [
      {
        serviceId: {
          type: mongoose.Schema.Types.ObjectId,
          ref: "Service",
          required: true,
        },
        selectedDate: { type: Date, required: true },
        timeSlotStart: { type: String, required: true },
        timeSlotEnd: { type: String, required: true },
        quantity: { type: Number, required: true, min: 1, default: 1 },
        priceAtBooking: { type: Number, required: true, min: 0 },
      },
    ],
    statusHistory: [
      {
        status: { type: String, required: true },
        timestamp: { type: Date, default: Date.now },
        reason: { type: String },
      },
    ],
    cancelledAt: { type: Date },
  },
  { timestamps: true },
);

const Booking = mongoose.model("Booking", bookingSchema);

export default Booking;

File: mobile/lib/features/bookings/data/models/booking_model.dart
class BookingItemModel {
  final String serviceId;
  final DateTime selectedDate;
  final String timeSlotStart;
  final String timeSlotEnd;
  final double priceAtBooking;

  const BookingItemModel({
    required this.serviceId,
    required this.selectedDate,
    required this.timeSlotStart,
    required this.timeSlotEnd,
    required this.priceAtBooking,
  });

  factory BookingItemModel.fromJson(Map<String, dynamic> json) {
    return BookingItemModel(
      serviceId: json['serviceId']?.toString() ?? '',
      selectedDate: DateTime.tryParse(json['selectedDate']?.toString() ?? '') ?? DateTime.fromMillisecondsSinceEpoch(0),
      timeSlotStart: json['timeSlotStart']?.toString() ?? '',
      timeSlotEnd: json['timeSlotEnd']?.toString() ?? '',
      priceAtBooking: (json['priceAtBooking'] as num?)?.toDouble() ?? 0,
    );
  }
}

class BookingModel {
  final String id;
  final String userId;
  final String status;
  final String paymentMethod;
  final String paymentStatus;
  final double totalAmount;
  final List<BookingItemModel> items;
  final DateTime createdAt;
  final DateTime updatedAt;

  const BookingModel({
    required this.id,
    required this.userId,
    required this.status,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.totalAmount,
    required this.items,
    required this.createdAt,
    required this.updatedAt,
  });

  factory BookingModel.fromJson(Map<String, dynamic> json) {
    final itemsJson = (json['items'] as List?) ?? const [];

    return BookingModel(
      id: json['_id']?.toString() ?? '',
      userId: json['userId']?.toString() ?? '',
      status: json['status']?.toString() ?? '',
      paymentMethod: json['paymentMethod']?.toString() ?? '',
      paymentStatus: json['paymentStatus']?.toString() ?? '',
      totalAmount: (json['totalAmount'] as num?)?.toDouble() ?? 0,
      items: itemsJson
          .whereType<Map>()
          .map((item) => BookingItemModel.fromJson(Map<String, dynamic>.from(item)))
          .toList(),
      createdAt: DateTime.tryParse(json['createdAt']?.toString() ?? '') ?? DateTime.fromMillisecondsSinceEpoch(0),
      updatedAt: DateTime.tryParse(json['updatedAt']?.toString() ?? '') ?? DateTime.fromMillisecondsSinceEpoch(0),
    );
  }
}
